### PR TITLE
Add Session type which can apply settings to requests created from it

### PR DESCRIPTION
**Pull request by sbstp**

### CODE REVIEW

**Comment by sbstp** [comment]
Review: @adamreichold

**Comment by sbstp** [comment]
Going to do a pass on the docs to make sure everything is still coherent.

**Comment by sbstp** [comment]
@adamreichold hey I was looking at the code and it seems like I missed something. Does rustls listen to the danger_accept_invalid_certs flag?

**Comment by adamreichold** [comment]
> hey I was looking at the code and it seems like I missed something. Does rustls listen to the danger_accept_invalid_certs flag?

It does not. It gets all of its configuration via the `ClientConfig` data type which is immutable because it is usually shared process-wide. Also note that those two flags do not exist when only the rustls backend is configured. As said in the original rustls MR, I see two options here: Implement API surface abstracting over native-tls and rustls covering the options we feel necessary to expose or align things by exposing `TlsConnector` directly so that library users can use that to configure things as they wish.

File: .travis.yml [hunk 1]
```diff
@@ -5,3 +5,5 @@ rust:
   - nightly
 script:
   - /bin/sh run-tests.sh
+  - rustup component add clippy
```
Comment (adamreichold) [inline .travis.yml]: I wonder if this could land separately?

File: src/request/mod.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    Error: From<V::Error>,
+{
+    let value = value.try_into()?;
+    headers.append(header, value);
+    Ok(())
+}
+
+/// Represents a request that's ready to be sent. You can inspect this object for information about the request.
+#[derive(Debug)]
+pub struct PreparedRequest<B> {
```
Comment (adamreichold) [inline src/request/mod.rs]: If the request builder is now called `RequestBuilder`, this could become `Request`?

Comment (sbstp) [reply src/request/mod.rs]: I didn't change the name of the RequestBuilder, it was already named this way. The PreparedRequest terminology comes from the `requests` library where a request ready to be sent is available in immutable form to inspect before sending.

File: .travis.yml (hunk 1 above)
Comment (sbstp) [reply .travis.yml]: Yes, I forgot to add cargo clippy --all-features anyway.

Comment (adamreichold) [reply .travis.yml]: We probably want both, a run with `--no-default-features` and a run with `--all-features`?

File: src/request/mod.rs (hunk 2 above)
Comment (adamreichold) [reply src/request/mod.rs]: Ah, sorry, seems I got confused here.

**Pull request merged**